contract_comm/evm.go

```diff
@@ -43,31 +41,95 @@ type InternalEVMHandler struct {
 	chain vm.ChainContext
 }
 
-func MakeStaticCall(registryId [32]byte, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, header *types.Header, state vm.StateDB) (uint64, error) {
-	return makeCallWithContractId(registryId, abi, funcName, args, returnObj, gas, nil, header, state, true)
+func MakeStaticCall(registryId common.Hash, abi abi.ABI, method string, args []interface{}, returnObj interface{}, gas uint64, header *types.Header, state vm.StateDB) (uint64, error) {
+	backend, err := createBackend(header, state)
+	if err != nil {
+		return 0, err
+	}
+
+	return Query(backend, registryId, &abi, method, args, returnObj, gas)
+}
+
+func Query(backend contracts.Backend, registryId common.Hash, abi *abi.ABI, method string, args []interface{}, returnObj interface{}, gas uint64) (uint64, error) {
+
+	contractAddress, err := resolveAddressForCall(backend, registryId, method)
+	if err != nil {
+		return 0, err
+	}
+
+	contract := contracts.NewContract(abi, contractAddress, contracts.SystemCaller)
+	gasLeft, err := contract.Query(contracts.QueryOpts{MaxGas: gas, Backend: backend}, returnObj, method, args...)
+
+	if err != nil {
+		log.Error("Error when invoking evm function", "err", err, "function", method, "address", contractAddress, "args", args, "gas", gas, "gasLeft", gasLeft)
+		return gasLeft, err
+	}
+
+	return gasLeft, nil
 }
 
-func MakeCall(registryId [32]byte, abi abi.ABI, funcName string, args []interface{}, returnObj interface{}, gas uint64, value *big.Int, header *types.Header, state vm.StateDB, finaliseState bool) (uint64, error) {
-	gasLeft, err := makeCallWithContractId(registryId, abi, funcName, args, returnObj, gas, value, header, state, false)
+func MakeCall(registryId common.Hash, abi abi.ABI, method string, args []interface{}, returnObj interface{}, gas uint64, value *big.Int, header *types.Header, state vm.StateDB, finaliseState bool) (uint64, error) {
+	backend, err := createBackend(header, state)
+	if err != nil {
+		return 0, err
+	}
+
+	gasLeft, err := Execute(backend, registryId, &abi, method, args, returnObj, gas, value)
+
 	if err == nil && finaliseState {
 		state.Finalise(true)
 	}
+
+	return gasLeft, err
+}
+
+func Execute(backend contracts.Backend, registryId common.Hash, abi *abi.ABI, method string, args []interface{}, returnObj interface{}, gas uint64, value *big.Int) (uint64, error) {
+	contractAddress, err := resolveAddressForCall(backend, registryId, method)
+	if err != nil {
+		return 0, err
+	}
+
+	contract := contracts.NewContract(abi, contractAddress, contracts.SystemCaller)
+	gasLeft, err := contract.Execute(contracts.ExecOpts{MaxGas: gas, Backend: backend, Value: value}, returnObj, method, args...)
+
+	if err != nil {
+		log.Error("Error when invoking evm function", "err", err, "function", method, "address", contractAddress, "args", args, "gas", gas, "gasLeft", gasLeft)
+		return gasLeft, err
+	}
+
 	return gasLeft, err
```

Comment:
Either make this `gasLeft, nil` or make the corresponding one in `Query` (line 68) also be `gasLeft, err` (in which case we don't need `return`'s inside the `if`'s.  Either way is fine, but nice to be consistent.